src/test/datascience/notebook/helper.ts

```diff
@@ -763,3 +767,37 @@ export async function hijackPrompt(
         clickButton: (text?: string) => clickButton.resolve(text || buttonToClick?.text)
     };
 }
+
+export async function asPromise<T>(
+    event: Event<T>,
+    timeout = env.uiKind === UIKind.Desktop ? 5000 : 15000
+): Promise<T> {
+    return new Promise<T>((resolve, reject) => {
+        const handle = setTimeout(() => {
+            // eslint-disable-next-line @typescript-eslint/no-use-before-define
+            sub.dispose();
+            reject(new Error('asPromise TIMEOUT reached'));
+        }, timeout);
+        const sub = event((e) => {
+            clearTimeout(handle);
+            sub.dispose();
+            resolve(e);
+        });
+    });
+}
+
+export async function waitForEvent<T>(
+    eventType: string,
+    debugAdapter: KernelDebugAdapter,
+    timeout = env.uiKind === UIKind.Desktop ? 5000 : 15000
+): Promise<T> {
+    setTimeout(() => {
+        throw new Error(`Debugging Timeout: ${eventType} message took too long to arrive`);
```

Comment:
You might want to clear the timeout to avoid throwing in the successful case. It might show up as an unhandled error in a test log.

Reply:
Also, you should return a rejected promise from waitForEvent when this triggers to make sure you get a test failure with the right error message.

Reply:
Yeah this exception isn't going to end up anywhere. You could just the already existing WaitForCondition

```suggestion
    return WaitForCondition(async () => {
       const msg: any = await asPromise(debugAdapter.onDidSendMessage);
       return msg.event == eventType;
    },
    timeout);
```

Reply:
That wouldn't get the message back

Reply:
Yeah sorry, the message is part of waitForCondition too. I just didn't type it in.

Reply:
Not sure what you mean @rchiodo, how can waitForCondition return a value?

Reply:
We need waitForEvent to return the event, only way I can think of is to capture the message in the function closure and return it after awaiting waitForCondition

Reply:
No you'd use waitForCondition in this function to wait for the message. Inside the function, just set a variable.

Like this:

```typescript
export async function waitForEvent<T>(
   eventType: string,
   debugAdapter: KernelDebugAdapter
   timeout = env.uiKind === UIKind.Desktop ? 5000 : 15000
): Promise<T> {
   let msg: T | undefined = undefined;
   await waitForCondition(async () => { 
    msg: any = await asPromise(debugAdapter.onDidSendMessage);
           msg = await asPromise(debugAdapter.onDidSendMessage);
	       return msg.event == eventType;
   }, `Debugging Timeout: ${eventType} message took too long to arrive`, timeout);
  return msg;
}
```

Reply:
That's what I was describing. But it's not the right approach because waitForCondition has a 10ms delay between loops so it's possible to miss an event (that might be possible even without the delay just because of async-ness) I implemented something simpler